src/library/arith_instance_manager.cpp

```diff
@@ -0,0 +1,489 @@
+        expr inst_type = mk_app(mk_constant(get_has_div_name(), {m_level}), m_type);
+        if (auto inst = m_tctx_ptr->mk_class_instance(inst_type)) {
+            m_div = mk_app(mk_constant(get_div_name(), {m_level}), m_type, *inst);
+            return m_div;
+        } else {
+            throw exception(sstream() << "cannot synthesize [has_div " << m_type << "]\n");
+        }
+    }
+}
+
+expr arith_instance_info::get_neg() {
+    if (!null(m_neg)) {
+        return m_neg;
+    } else {
+        lean_assert(m_tctx_ptr);
+        expr inst_type = mk_app(mk_constant(get_has_neg_name(), {m_level}), m_type);
+        if (auto inst = m_tctx_ptr->mk_class_instance(inst_type)) {
+            m_neg = mk_app(mk_constant(get_neg_name(), {m_level}), m_type, *inst);
+            return m_neg;
+        } else {
+            throw exception(sstream() << "cannot synthesize [has_neg " << m_type << "]\n");
+        }
+    }
+}
+
+expr arith_instance_info::get_eq() { return mk_app(mk_constant(get_eq_name(), {m_level}), m_type); }
+
+expr arith_instance_info::get_lt() {
+    if (!null(m_lt)) {
+        return m_lt;
+    } else {
+        lean_assert(m_tctx_ptr);
+        expr inst_type = mk_app(mk_constant(get_has_lt_name(), {m_level}), m_type);
+        if (auto inst = m_tctx_ptr->mk_class_instance(inst_type)) {
+            m_lt = mk_app(mk_constant(get_lt_name(), {m_level}), m_type, *inst);
+            return m_lt;
+        } else {
+            throw exception(sstream() << "cannot synthesize [has_lt " << m_type << "]\n");
+        }
+    }
+}
+
+expr arith_instance_info::get_le() {
+    if (!null(m_le)) {
+        return m_le;
+    } else {
+        lean_assert(m_tctx_ptr);
+        expr inst_type = mk_app(mk_constant(get_has_le_name(), {m_level}), m_type);
+        if (auto inst = m_tctx_ptr->mk_class_instance(inst_type)) {
+            m_le = mk_app(mk_constant(get_le_name(), {m_level}), m_type, *inst);
+            return m_le;
+        } else {
+            throw exception(sstream() << "cannot synthesize [has_le " << m_type << "]\n");
+        }
+    }
+}
+
+// Entry points
+arith_instance_info & get_arith_instance_info_for(concrete_arith_type type) {
+    switch (type) {
+    case concrete_arith_type::NAT:  return *g_nat_instance_info;
+    case concrete_arith_type::INT:  return *g_int_instance_info;
+    case concrete_arith_type::REAL: return *g_real_instance_info;
+    }
+    lean_unreachable();
+}
+
+arith_instance_info & get_arith_instance_info_for_nat() {
+    return *g_nat_instance_info;
+}
+
+arith_instance_info & get_arith_instance_info_for_int() {
+    return *g_int_instance_info;
+}
+
+arith_instance_info & get_arith_instance_info_for_real() {
+    return *g_real_instance_info;
+}
+
+arith_instance_info & get_arith_instance_info_for(type_context & tctx, expr const & type) {
+    expr_struct_map<arith_instance_info> & cache = get_arith_instance_info_cache_for(tctx);
+    auto it = cache.find(type);
+    if (it != cache.end()) {
```

Comment:
The approach used for caching looks incorrect, and may produce incorrect results.
Consider the following corner case scenario. 
Invoke `get_arith_instance_infor_for(tctx, A)` where the local context in `tctx` is 
`(A : Type) [comm_ring A] ...`
Now, we invoke `get_arith_instance_info_for(tctx, A)` again with the same `A` and same environment, but the rest of the local context is different (e.g., `(A : Type) [field A] ...`). In this scenario, the cache would contain incorrect results.
I realize this is not a common scenario, but it may generate problems in the future.
To fix this problem we have to use an approach similar to the one used in `type_context`.
It is more expensive, since we have to check whether two different local contexts are compatible or not. when we are accessing the cache.